### PR TITLE
Implement client feature for Discovery API

### LINKED ISSUE #204: Implement client feature for Discovery API
Support getting the details of the supported oras extension from the extension API.

Spec: https://github.com/oras-project/artifacts-spec/blob/main/manifest-referrers-api.md#api-discovery

### PR DESCRIPTION
**Pull request by m5i-work**
Resolves #204 

Signed-off-by: Wei Meng <wemeng@microsoft.com>

### CODE REVIEW

File: registry/remote/repository.go [hunk 1]
```diff
@@ -437,6 +437,58 @@ func filterReferrers(refs []artifactspec.Descriptor, artifactType string) []arti
...
+	}
+	lr := limitReader(resp.Body, r.MaxMetadataBytes)
+	if err := json.NewDecoder(lr).Decode(&discoveryResult); err != nil {
+		return nil, fmt.Errorf("%s %q: failed to decode response: %w", resp.Request.Method, resp.Request.URL, err)
+	}
+	return discoveryResult.Extensions, nil
+}
+
+// DiscoverExtension finds the first extension with give name in all supported extensions in current repository.
+// (nil, errdef.ErrNotFound) is returned if no such extension found.
```
[the remark this answers is not in the document]
Comment (Wwwsylvia) [reply registry/remote/repository.go:477]: Like this: https://github.com/oras-project/oras-go/blob/main/content/memory/memory.go#L76

File: registry/remote/repository.go [hunk 2]
```diff
@@ -437,6 +437,58 @@ func filterReferrers(refs []artifactspec.Descriptor, artifactType string) []arti
...
+		Extensions []Extension `json:"extensions"`
+	}
+	lr := limitReader(resp.Body, r.MaxMetadataBytes)
+	if err := json.NewDecoder(lr).Decode(&discoveryResult); err != nil {
+		return nil, fmt.Errorf("%s %q: failed to decode response: %w", resp.Request.Method, resp.Request.URL, err)
+	}
+	return discoveryResult.Extensions, nil
+}
+
+// DiscoverExtension finds the first extension with give name in all supported extensions in current repository.
```
[the remark this answers is not in the document]
Comment (Wwwsylvia) [reply registry/remote/repository.go]: The recommend length is `80`.

File: registry/remote/repository.go [hunk 3]
```diff
@@ -437,6 +437,59 @@ func filterReferrers(refs []artifactspec.Descriptor, artifactType string) []arti
...
+type Extension struct {
+	Name        string   `json:"name"`
+	URL         string   `json:"url"`
+	Description *string  `json:"description,omitempty"`
+	Endpoints   []string `json:"endpoints"`
+}
+
+// DiscoverExtensions lists all supported extensions in current repository.
+// Reference: https://github.com/oras-project/artifacts-spec/blob/main/manifest-referrers-api.md#api-discovery
+func (r *Repository) DiscoverExtensions(ctx context.Context) ([]Extension, error) {
```
Comment (shizhMSFT) [inline registry/remote/repository.go:451]: Does this API support pagination?

Related: https://github.com/opencontainers/distribution-spec/tree/main/extensions#pagination

File: registry/remote/repository.go [hunk 4]
```diff
@@ -437,6 +437,59 @@ func filterReferrers(refs []artifactspec.Descriptor, artifactType string) []arti
...
 }
 
+// Extension provides information on extensions provided on a Registry.
+// Reference: https://github.com/opencontainers/distribution-spec/blob/main/extensions/_oci.md#extensions-property-descriptions
+type Extension struct {
+	Name        string   `json:"name"`
+	URL         string   `json:"url"`
+	Description *string  `json:"description,omitempty"`
+	Endpoints   []string `json:"endpoints"`
+}
```
Comment (shizhMSFT) [inline registry/remote/repository.go:447]: The go struct from the spec may be reused.

Reference: https://github.com/opencontainers/distribution-spec/blob/main/specs-go/v1/extensions/ext.go

File: registry/remote/repository.go [hunk 5]
```diff
@@ -437,6 +437,59 @@ func filterReferrers(refs []artifactspec.Descriptor, artifactType string) []arti
...
+	if err != nil {
+		return Extension{}, fmt.Errorf("Failed to discover extensions: %w", err)
+	}
+	for _, ext := range extensions {
+		if ext.Name == name {
+			return ext, nil
+		}
+	}
+	return Extension{}, errdef.ErrNotFound
+}
```
Comment (shizhMSFT) [inline registry/remote/repository.go:491]: What's the use case for this API? We might want to remove it if we don't have a clear use case.

**Pull request merged**